proposals/type-applications-in-patterns.rst

```diff
@@ -0,0 +1,170 @@
+.. sectnum::
+.. contents::
+
+We have ``TypeApplications`` in terms. This proposal brings them to patterns in a way that preserves analogy to type signatures.
+
+
+Motivation
+------------
+
+``TypeApplications`` are a convenient and natural way to specifying types of polymorphic functions. Consider::
+
+ data Foo a where MkFoo :: forall a. a -> Foo a
+ 
+With ``TypeApplications``, I can replace the somewhat clumsy ``MkFoo (x :: ty)`` with ``MkFoo @ty x``. Seen this way,
+explicit type applications are merely alternative syntax for type signatures.
+
+At the moment, this only works in terms, but not in patterns: We can use type signatures in patterns
+(if ``PatternSignatures`` or ``ScopedTypeVariables`` are enabled), but not type applications. Given the strong
+relation between these syntactic forms, this is odd – why can I write::
+
+    foo (MkFoo (x :: ty)) = …
+   
+but not::
+
+    foo (MkFoo @ty x) = …
+
+This proposal fills this gap, by allowing type applications in pattern syntax and specifying it to behave
+“just like type signatures”.
+
+The intention of the following specification is that the following holds: For a constructor with type ``C :: forall a. a -> …`` the meaning of ``C @ty x`` should coincide with the existing form ``C (x :: ty)``. 
+
+Proposed Change Specification
+-----------------------------
+
+When both ``TypeApplications`` and ``PatternSignatures`` are enabled, then type application syntax is
+available in patterns. 
+
+If ``ScopedTypeVariables`` is enabled, then type applications and type signatures in pattern may mention type variables.
+
+A type variable mentioned in a pattern (in a type signature or a type application) that is not already in scope is brought into scope. The scope spans the pattern and the corresponding right-hand side.
+
+Typechecking a type application in a pattern follows the same rules as typechecking a pattern signature. This proposal does not propose any changes to the mechanism. In particular
+
+* type variables in type applications in patterns can only bind type variables, just as it is the case for type variables in pattern signatures (unless `Proposal #128 <https://github.com/ghc-proposals/ghc-proposals/pull/128>`_ gets accepted first).
+* patterns are type-checked from left to right. This means when type-checking the type application in a pattern ``(MkT @ty)``, type equatlities from matches further to the left, as well as the type equalities from the context of ``MkT`` itself are in scope, but not type equalities from constructors further to the right (see below for an example).
+
+A more formal description of these rules can be found in `“Type variable in pattern” by Richard Eisenberg, Joachim Breitner  and Simon Peyton Jones <https://arxiv.org/abs/1806.03476>`_.
+
+An underscore in a type signature or type application in a pattern is not treated as a hole.
+
+Examples
+--------
+
+Here is an example (taken from _#15050 <https://ghc.haskell.org/trac/ghc/ticket/15050#comment:10>_)::
```

Comment:
Please point out explicitly in this example that you are supplying only the first of two type arguments; so you are giving a type that matches `b`, but not `a`, because the data type declaration says `forall b a...`

Reply:
✓